### PR TITLE
Add extendeddaemonset-check helper

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

Add extendeddaemonset-check helper

* add new cmd binary `check-eds` which can check when a EDS update is finished.
* add new `app-example` chart: example of an application using the `ExtendedDaemonset`.
* update ci to deliver the `datadog/extendeddaemonset-check` container image.

### Motivation

Helm3 is not able to know when the deployment of an application deployed with an`ExtendedDaemonset` is done.
The new `datadog/extendeddaemonset-check` can be used as helm3 test.

### Additional Notes

N/A

### Describe your test plan

To test it, please follow the instruction in the `charts/app-example/README.md`

### CODE REVIEW

File: chart/app-example/templates/extendeddaemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        - name: {{ .Chart.Name }}
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          resources:
+            {{- toYaml .Values.resources | nindent 12 }}
+      tolerations:
+      - operator: Exists
+      {{- toYaml . | nindent 6 }}
```
Comment (L3n41c) [inline chart/app-example/templates/extendeddaemonset.yaml:29]: I couldn’t find any englobing and not yet closed `with` statement.
This `.`  will pull all the variables, won’t it ?

Anyhow, the `- operator: Exists` toleration is already tolerating all the taints. So, there is no value in allowing the end-user to add some extra useless toleration.

```suggestion
```

File: chart/app-example/values.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  #   memory: 128Mi
+
+autoscaling:
+  enabled: false
+  minReplicas: 1
+  maxReplicas: 100
+  targetCPUUtilizationPercentage: 80
+  # targetMemoryUtilizationPercentage: 80
+
+tolerations:
```
Comment (L3n41c) [inline chart/app-example/values.yaml:75]: The EDS already has an hardcoded “match all” toleration.
If there’s no way to remove this “match all” toleration, it’s useless to be able to add some others.
```suggestion
```

File: cmd/check-eds/upgrade/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+func (o *Options) Run() error {
+	o.printOut("start checking deployment state")
+
+	checkUpgradeDown := func() (bool, error) {
+		eds := &v1alpha1.ExtendedDaemonSet{}
+		err := o.client.Get(context.TODO(), client.ObjectKey{Namespace: o.userNamespace, Name: o.userExtendedDaemonSetName}, eds)
+		if err != nil && errors.IsNotFound(err) {
+			return false, fmt.Errorf("ExtendedDaemonSet %s/%s not found", o.userNamespace, o.userExtendedDaemonSetName)
+		} else if err != nil {
+			return false, fmt.Errorf("unable to get ExtendedDaemonSet, err: %v", err)
```
Comment (L3n41c) [inline cmd/check-eds/upgrade/upgrade.go:136]:
```suggestion
			return false, fmt.Errorf("unable to get ExtendedDaemonSet, err: %w", err)
```

File: chart/app-example/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# APP Example
+
+## Introduction
+
+This chart allows you to deploy a "dummy" application that used the ExtendedDaemonset.
```
Comment (L3n41c) [inline chart/app-example/README.md:5]:
```suggestion
This chart allows you to deploy a "dummy" application that uses the ExtendedDaemonset.
```

File: chart/app-example/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+thanks to this chart you can test several ExtendedDaemonset features, such as:
```
Comment (L3n41c) [inline chart/app-example/README.md:7]:
```suggestion
Thanks to this chart you can test several ExtendedDaemonset features, such as:
```

File: chart/app-example/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+* The Canary deployment strategy
+* The `extendeddaemonset-check` util pod with the `helm test` command.
+* The possibility to override a `Pod` resources for a specific `Node` thanks to an
+  `ExtendedDaemonsetSettings` resource.
+
+## Deployment
+
+Before deploying this chart, you should have deployed the `ExtendedDaemonset` Controller with the following.
```
Comment (L3n41c) [inline chart/app-example/README.md:16]:
```suggestion
Before deploying this chart, you should have deployed the `ExtendedDaemonset` Controller with the following
```

**Review by L3n41c — COMMENTED** [review]
Just left a few cosmetics comments. Mostly about dead or useless things.

File: chart/app-example/templates/extendeddaemonset.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    {{- include "app-example.labels" . | nindent 4 }}
+spec:
+  strategy:
+    {{- toYaml .Values.strategy | nindent 4 }}
+  template:
+    spec: 
+      {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply chart/app-example/templates/extendeddaemonset.yaml:15]: `with` replace `if` and also change the current scope by the value tested.
More [docs](https://helm.sh/docs/chart_template_guide/control_structures/#modifying-scope-using-with)

File: chart/app-example/templates/tests/role.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{{- if .Values.rbac.create -}}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: "{{ include "app-example.fullname" . }}-test"
+  labels:
+    {{- include "app-example.labels" . | nindent 4 }}
+  annotations:
+    "helm.sh/hook": test
```
[the remark this answers is not in the document]
Comment (clamoriniere) [reply chart/app-example/templates/tests/role.yaml:9]: this Role is only created if the user run the command: `helm test`

File: chart/app-example/templates/extendeddaemonset.yaml (hunk 1 above)
Comment (clamoriniere) [reply chart/app-example/templates/extendeddaemonset.yaml:29]: good catch

**Review by L3n41c — APPROVED** [review]

**Pull request merged**